### ISSUE TITLE
High false positive

### ISSUE THREAD

**Issue by UmairDL** [dscripka/openWakeWord#255]
I am using this training_models.ipynb file. However facing high false positive rate issue, following are my parameters:
- positive samples : 30k (synthetically generated using vits, piper and waveglow)
- negative samples: 393867
- Tried different clip size as well 1s, 2s and 3s
- Tried different model dimensions : 32, 64,128
- Epochs: 500
Not sure, how can i decrease the false positive. Would really appreciate your suggestion.

**Comment by sangheonEN** [comment]
Can I ask you a question that is a little off topic?

[Question 1]
I'm curious about how you built the learning environment for waveglow. I built it using the repository https://github.com/dscripka/synthetic_speech_dataset_generation, but as you can see in https://github.com/dscripka/synthetic_speech_dataset_generation/issues/4, 10 samples were generated without sound. How can I solve this?

Ultimately, what I want is to get positive samples using the voices of various people that have been trained.

[Question 2]
I'm curious as to how you constructed the negative sample.

Did you build it as shown here? (https://github.com/dscripka/openWakeWord/blob/main/docs/models/hey_jarvis.md)

And can you share the source code of training_models.ipynb?

I know you're busy, but I would really appreciate it if you could reply.

**Comment by UmairDL** [comment]
Hi @sangheonEN, 

For your [Question 1]:
Iam running this on linux environment.
I created virtual environment and for that i used python 3.8.20 version and installed the following libraries:
Package                   Version
------------------------- -----------
aiohappyeyeballs          2.4.4
aiohttp                   3.10.11
aiosignal                 1.3.1
alembic                   1.14.0
annotated-types           0.7.0
async-timeout             5.0.1
attrs                     24.3.0
audioread                 3.0.1
babel                     2.16.0
BentoML                   0.12.1
bibtexparser              2.0.0b8
blinker                   1.8.2
boto3                     1.35.96
botocore                  1.35.96
Cerberus                  1.3.7
certifi                   2024.12.14
cffi                      1.17.1
chardet                   5.2.0
charset-normalizer        3.4.1
click                     8.1.8
clldutils                 3.24.0
colorama                  0.4.6
colorlog                  6.9.0
configparser              7.1.0
csvw                      3.5.1
decorator                 5.1.1
deepmerge                 2.0
dependency-injector       4.45.0
dlinfo                    1.2.1
docker                    5.0.3
filelock                  3.16.1
Flask                     3.0.3
frozenlist                1.5.0
fsspec                    2024.12.0
grpcio                    1.69.0
gunicorn                  23.0.0
humanfriendly             10.0
idna                      3.10
importlib_metadata        8.5.0
importlib_resources       6.4.5
inflect                   6.2.0
isodate                   0.7.2
itsdangerous              2.2.0
Jinja2                    3.1.5
jmespath                  1.0.1
joblib                    1.4.2
jsonschema                4.23.0
jsonschema-specifications 2023.12.1
language-tags             1.2.0
librosa                   0.9.2
llvmlite                  0.41.1
lxml                      5.3.0
Mako                      1.3.8
Markdown                  3.7
MarkupSafe                2.1.5
mpmath                    1.3.0
multidict                 6.1.0
networkx                  3.1
numba                     0.58.1
numpy                     1.23.3
nvidia-cublas-cu12        12.1.3.1
nvidia-cuda-cupti-cu12    12.1.105
nvidia-cuda-nvrtc-cu12    12.1.105
nvidia-cuda-runtime-cu12  12.1.105
nvidia-cudnn-cu12         8.9.2.26
nvidia-cufft-cu12         11.0.2.54
nvidia-curand-cu12        10.3.2.106
nvidia-cusolver-cu12      11.4.5.107
nvidia-cusparse-cu12      12.1.0.106
nvidia-nccl-cu12          2.19.3
nvidia-nvjitlink-cu12     12.6.85
nvidia-nvtx-cu12          12.1.105
packaging                 24.2
phonemizer                3.3.0
pip                       22.0.2
pkgutil_resolve_name      1.3.10
platformdirs              4.3.6
pooch                     1.8.2
prometheus_client         0.21.1
propcache                 0.2.0
protobuf                  3.19.1
psutil                    6.1.1
pycparser                 2.22
pydantic                  2.10.4
pydantic_core             2.27.2
pylatexenc                2.10
pyparsing                 3.1.4
python-dateutil           2.9.0.post0
python-json-logger        3.2.1
pytz                      2024.2
rdflib                    7.1.1
referencing               0.35.1
regex                     2024.11.6
requests                  2.32.3
resampy                   0.4.3
rfc3986                   1.5.0
rpds-py                   0.20.1
ruamel.yaml               0.18.10
ruamel.yaml.clib          0.2.8
s3transfer                0.10.4
schema                    0.7.7
scikit-learn              1.3.2
scipy                     1.10.1
segments                  2.2.1
setuptools                59.6.0
six                       1.17.0
soundfile                 0.13.0
SQLAlchemy                1.3.24
SQLAlchemy-Utils          0.36.5
sympy                     1.13.3
tabulate                  0.9.0
threadpoolctl             3.5.0
torch                     2.2.2
torchaudio                2.2.2
tqdm                      4.67.1
triton                    2.2.0
typing_extensions         4.12.2
Unidecode                 1.3.6
uritemplate               4.1.1
urllib3                   1.25.11
websocket-client          1.8.0
Werkzeug                  3.0.6
wheel                     0.37.1
yarl                      1.15.2
zipp                      3.20.2

**Comment by UmairDL** [comment]
@sangheonEN AND for question 2:

For negative samples i just downloaded the files and extracted them into my working directory,  those directories contains .wav format audio files. 

AND the code iam using is exactly the same as provided in github repo i.e., [link](https://github.com/dscripka/openWakeWord/blob/main/notebooks/training_models.ipynb)

**Comment by sangheonEN** [comment]
Thanks for the reply.

Can you tell me which file you downloaded?

https://github.com/dscripka/openWakeWord/blob/main/docs/models/hey_jarvis.md

Of the 4 methods described here, only method 2 is available for download.

2. ~10,000 hours from the Common Voice 11 dataset, representing multiple languages

https://commonvoice.mozilla.org/en/datasets

**Comment by UmairDL** [comment]
I was also able to use only these two,
1. ~10,000 hours from the [Common Voice 11 dataset](https://commonvoice.mozilla.org/en/datasets), representing multiple languages
2. ~1,000 hours of music from the [Free Music Archive dataset](https://github.com/mdeff/fma)

**Comment by sangheonEN** [comment]
Oh, so you're saying that you sampled from that dataset(common voice 11, free music archive) and constructed a negative sample?

**Comment by UmairDL** [comment]
correct

**Comment by sangheonEN** [comment]
wake word : hey thomas
positive samples : 70k (synthetically generated using vits, piper)
- VITS : python3 generate_clips.py --model VITS --enable_gpu --text "hey tho ma su" --N 5000 --max_per_speaker 1 --output_dir ./generated_clips_VITS
- not augment_clips
negative samples: 170,827
clip size : 2s
Tried different model dimensions : 32
steps: 50000

This is the result of testing with a voice file saying "Hey Thomas" twice in chunks 40~60 and 80~100, as shown in the graph. The results aren't very good... It seems like the model didn't learn properly.

![Image](https://github.com/user-attachments/assets/e6784c6e-9db2-4b12-834b-3dda4e0ba53e)

**Comment by sangheonEN** [comment]
I evaluated false positives with the code below. I used the validation_set_features.npy validation data provided by the corresponding git.

However, the result is miserable. Why do so many false positives occur? For my training, please refer to https://github.com/sangheonEN/openWakeWord/blob/main/notebooks/training_models.ipynb.

![Image](https://github.com/user-attachments/assets/f69eee62-fb60-4ecc-b832-657ba47a8631)

Could it be because I didn't apply the loss term for false positive? However, the result of training and evaluating the model with the applied code https://github.com/sangheonEN/openWakeWord/blob/main/lee_custom_train.py is terrible, as shown in the graph below, and the prediction value doesn't come out properly. What did I do wrong?

![Image](https://github.com/user-attachments/assets/57f886f5-6bc1-40f2-9446-d569e8b687fb)


```python
import numpy as np
import onnxruntime
import matplotlib.pyplot as plt
from tqdm import tqdm

# 1. Load features from file (shape: [N, 28, 96])
npy_path = "/home/data/openwakeword/validation_set_features.npy"
features = np.load(npy_path)

# 2. Load ONNX model
onnx_model_path = "/home/openWakeWord/models/hey_thomas20250409.onnx"
session = onnxruntime.InferenceSession(onnx_model_path)
input_name = session.get_inputs()[0].name

# 3. 슬라이딩 윈도우 적용해서 예측
window_size = 16
scores = []
for i in tqdm(range(0, features.shape[0] - window_size)):
    window = features[i:i + window_size].astype(np.float32).reshape(1, window_size, 96)
    output = session.run(None, {input_name: window})
    scores.append(float(output[0][0][0]))  # output shape: (1, 1)

# 4. 결과 시각화
plt.figure(figsize=(12, 4))
plt.plot(scores, label="False Positive Confidence")
plt.axhline(0.5, color='red', linestyle='--', label="Threshold = 0.5")
plt.xlabel("Window Index")
plt.ylabel("Confidence")
plt.title("False Positive Scores on Validation Features")
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.savefig("false_positive_eval_from_framewise.png")
plt.show()
```
